src/library/scala/concurrent/Future.scala

```diff
@@ -513,80 +527,116 @@ object Future {
     classOf[Unit]    -> classOf[scala.runtime.BoxedUnit]
   )
 
+  private[this] final val _cachedId: AnyRef => AnyRef = Predef.identity _
+
+  private[concurrent] final def id[T]: T => T = _cachedId.asInstanceOf[T => T]
+
+  private[concurrent] final val collectFailed =
+    (t: Any) => throw new NoSuchElementException("Future.collect partial function is not defined at: " + t) with NoStackTrace
+
+  private[concurrent] final val filterFailure =
+    Failure[Nothing](new NoSuchElementException("Future.filter predicate is not satisfied") with NoStackTrace)
```

Comment:
@NthPortal Given our conversation about addSuppressed, how should we deal with this? By not reusing cached exceptions? I fear ending up with exceptions with tons of addSuppressed building up over time. Could also leak info between parts of the program.

Reply:
yeah, I think I would agree that we shouldn't cache the exception. It still should be fairly lightweight to construct though.

Reply:
The only other option I can think of is a custom exception (that doesn't extend `NoSuchElementException`) which disables suppression and stack traces